pkg/volume/util/operationexecutor/operation_executor.go

```diff
@@ -0,0 +1,763 @@
+	actualStateOfWorld ActualStateOfWorldMounterUpdater) (func() error, error) {
+	// Get mountable plugin
+	mountableVolumePlugin, err :=
+		oe.volumePluginMgr.FindPluginByName(volumeToUnmount.PluginName)
+	if err != nil || mountableVolumePlugin == nil {
+		return nil, fmt.Errorf(
+			"UnmountVolume.FindPluginByName failed for volume %q (volume.spec.Name: %q) pod %q (UID: %q) err=%v",
+			volumeToUnmount.VolumeName,
+			volumeToUnmount.OuterVolumeSpecName,
+			volumeToUnmount.PodName,
+			volumeToUnmount.PodUID,
+			err)
+	}
+
+	volumeUnmounter, newUnmounterErr := mountableVolumePlugin.NewUnmounter(
+		volumeToUnmount.InnerVolumeSpecName, volumeToUnmount.PodUID)
+	if newUnmounterErr != nil {
+		return nil, fmt.Errorf(
+			"UnmountVolume.NewUnmounter failed for volume %q (volume.spec.Name: %q) pod %q (UID: %q) err=%v",
+			volumeToUnmount.VolumeName,
+			volumeToUnmount.OuterVolumeSpecName,
+			volumeToUnmount.PodName,
+			volumeToUnmount.PodUID,
+			newUnmounterErr)
+	}
+
+	return func() error {
+		// Execute unmount
+		unmountErr := volumeUnmounter.TearDown()
+
+		if unmountErr != nil {
+			// On failure, just log and exit. The controller will retry
+			glog.Errorf(
+				"UnmountVolume.TearDown failed for volume %q (volume.spec.Name: %q) pod %q (UID: %q) with: %v",
+				volumeToUnmount.VolumeName,
+				volumeToUnmount.OuterVolumeSpecName,
+				volumeToUnmount.PodName,
+				volumeToUnmount.PodUID,
+				unmountErr)
+			return unmountErr
+		}
+
+		glog.Infof(
+			"UnmountVolume.TearDown succeeded for volume %q (volume.spec.Name: %q) pod %q (UID: %q).",
+			volumeToUnmount.VolumeName,
+			volumeToUnmount.OuterVolumeSpecName,
+			volumeToUnmount.PodName,
+			volumeToUnmount.PodUID)
+
+		// Update actual state of world
+		actualStateOfWorld.MarkVolumeAsUnmounted(
+			volumeToUnmount.PodName, volumeToUnmount.VolumeName)
+
+		return nil
+	}, nil
+}
+
+func (oe *operationExecutor) generateUnmountDeviceFunc(
+	deviceToDetach AttachedVolume,
+	actualStateOfWorld ActualStateOfWorldMounterUpdater) (func() error, error) {
+	// Get attacher plugin
+	attachableVolumePlugin, err :=
+		oe.volumePluginMgr.FindAttachablePluginBySpec(deviceToDetach.VolumeSpec)
+	if err != nil || attachableVolumePlugin == nil {
+		return nil, fmt.Errorf(
+			"UnmountDevice.FindAttachablePluginBySpec failed for volume %q (spec.Name: %q) with: %v",
+			deviceToDetach.VolumeName,
+			deviceToDetach.VolumeSpec.Name(),
+			err)
+	}
+
+	volumeDetacher, err := attachableVolumePlugin.NewDetacher()
+	if err != nil {
+		return nil, fmt.Errorf(
+			"UnmountDevice.NewDetacher failed for volume %q (spec.Name: %q) with: %v",
+			deviceToDetach.VolumeName,
+			deviceToDetach.VolumeSpec.Name(),
+			err)
+	}
+
+	volumeAttacher, err := attachableVolumePlugin.NewAttacher()
+	if err != nil {
+		return nil, fmt.Errorf(
+			"UnmountDevice.NewAttacher failed for volume %q (spec.Name: %q) with: %v",
+			deviceToDetach.VolumeName,
+			deviceToDetach.VolumeSpec.Name(),
+			err)
+	}
+
```

Comment:
Added more detailed comments to the Interface methods describing what each method does.